### PR TITLE
feat: Automatic attention slicing when slice size is set to 0

### PR DESCRIPTION
**Pull request by sagar-a16z**
Add support for automatic attention slicing based on the huggingface diffusers implementation https://github.com/huggingface/diffusers/blob/91925fbb761d944d54271660c4c3cffee55798fa/examples/community/stable_diffusion_mega.py#L96-L113

### CODE REVIEW

File: examples/stable-diffusion/main.rs [hunk 1]
```diff
@@ -222,6 +222,8 @@ fn run(args: Args) -> anyhow::Result<()> {
     tch::maybe_init_cuda();
     println!("Cuda available: {}", tch::Cuda::is_available());
     println!("Cudnn available: {}", tch::Cuda::cudnn_is_available());
+    println!("MPS available: {}", tch::utils::has());
```
Comment (LaurentMazare) [inline examples/stable-diffusion/main.rs:225]: Is this expected, maybe you meant `has_mps`?

File: src/models/unet_2d.rs [hunk 2]
```diff
@@ -103,6 +103,13 @@ impl UNet2DConditionModel {
             .map(|i| {
                 let BlockConfig { out_channels, use_cross_attn, attention_head_dim } =
                     config.blocks[i];
+
+                // Enable automatic attention slicing if the config sliced_attention_size is set to 0.
+                let sliced_attention_size = match config.sliced_attention_size {
+                    Some(s) => if s == 0 { Some(attention_head_dim / 2) } else { None },
```
Comment (LaurentMazare) [inline src/models/unet_2d.rs:109]: Using `Some(0)` in the pattern matching would feel a bit more idiomatic, any reason not to do so?

Comment (sagar-a16z) [reply src/models/unet_2d.rs:109]: No reason, updating it!

**Review by LaurentMazare — APPROVED** [review]

**Comment by LaurentMazare** [comment]
Looks good, thanks! Could you just give a bit more details about the testing that you did? (typically if you can generate images with/without the features and check that they compare well, that's great)

**Comment by sagar-a16z** [comment]
I tested it with the following command:
```
cargo run --target=aarch64-apple-darwin --example stable-diffusion --features clap -- --prompt "A very rusty robot holding a fire torch." --sliced-attention-size 0 --cpu unet --cpu clip
 ```
 
 Here's the output: 
image

Here's the outuput without `sliced-attention-size 0`
```
cargo run --target=aarch64-apple-darwin --example stable-diffusion --features clap -- --prompt "A very rusty robot holding a fire torch." --cpu unet --cpu clip
 ```
 output: 
image

The performance difference on my m1 mac is staggering...The first run finished in less then 2 minutes. Without attention slicing it takes over 10 minutes.

**Comment by LaurentMazare** [comment]
Neat, that's some impressive speedup! Thanks for the PR!

**Pull request merged**